### ISSUE TITLE
dm critical when not available

### ISSUE THREAD

**Issue by tesink** [glensc/nagios-plugin-check_raid#142]
Hi Glen,

Thanks for the great plugin. Since version 4.0.2 we have one issue; the plugin is issuing critical exit messages when there are no dm devices:

CRITICAL: dm:[No devices to check]: megacli:[Volumes(0): : Devices(4): 00,01,02,03=Online]

Please let me know if you need more details on this issue.

**Comment by glensc** [comment]
#138

**Comment by tersmitten** [comment]
Just upgraded to `4.0.3` now I'm having trouble with `dm:[No devices to check]`:

```
UNKNOWN: dm:[No devices to check]; hpacucli:[Smart Array P410[OK]: Array A(OK)[LUN1:OK]]
```

I found this `--noraid=OK`. Is that the (only) way to get thinks working again?

**Comment by glensc** [comment]
given you need just one plugin to check, `-p hpacucli` from the output

**Comment by tersmitten** [comment]
Ok, but why does it check `dm` by default?

**Comment by prupert** [comment]
IMHO it should **only do `dm` checks if there actually are arrays configured in `dm`**.

**Comment by glensc** [comment]
because this is by design. read README.md and related bugreports linked from this issue, feel free to help and send your version of PR.

**Comment by glensc** [comment]
@prupert read the related bugreports

**Comment by calestyo** [comment]
Is there a way to set --noraid=OK **just** for a specific plugin (i.e. dm)?
The thing is, most people will have dmsetup installed even if not using MD at all, simply because DM is much more than just RAID/MD (dm-crypt, dm-integrity, cache and so on).

OTOH, I'd like to avoid using -p, cause it seems nice that check_raid simply works out-of-the box on every node, when the respective raid tools are installed.

**Comment by glensc** [comment]
the specific case is that dm is not properly implemented, i believe link  to #138 explains it. as you see no one has implemented anything since that. due lack of time, lack of interest or lack of knowledge.

**Comment by calestyo** [comment]
Ah, I see =)

Maybe the following could be done:
```
# dmsetup status
system: 0 1965027328 crypt 
```
status gives a list of all DM devices and their type, with the types given in the manpage in the TABLE FORMAT section.

So check_raid could first run dmsetup status, and grep per default for types which are actually something like a RAID, e.g.:
mirror and raid

With and plugin option one could allow people to override which types they look for, including "", i.e. ignore anything.

Unless a device in the list remains after that matching, dm would simply be ignored in check_raid

**Comment by glensc** [comment]
technical implementation was discussed somewhere in tickets/pr's. if really interested try reading and navigating between them yourself (likely there are cross references). i recall top of my head that why the detect was not done was related to sudo invocation. haven't got time to deal with it any more.

**Comment by glensc** [comment]
started work on [dm-detect](https://github.com/glensc/nagios-plugin-check_raid/tree/dm-detect( branch:

however  there's technical problem with sudo rules:
- current test to see if plugin should be enabled requires sudo access
- sudo rules are generated based whether plugin is enabled or not

so in the end plugin never be enabled for non-root because sudo rules are not generated.

so it's kind of chicken-egg problem. it needs to be solved first.

**Comment by glensc** [comment]
will be solved in #172
